galley/pkg/mcp/client/client.go

```diff
@@ -0,0 +1,310 @@
+// for the following:
+//
+// - Maintaining the bidirectional grpc stream with the server. The stream will be reestablished
+//   on transient network failures. The provided grpc connection (mcpClient) is assumed to handle
+//   (re)dialing the server.
+//
+// - Handling all aspects of the MCP exchange for the supported message types, e.g. request/response,
+//   ACK/NACK, nonce, versioning,
+//
+// - Decoding the received configuration updates and providing them to the user via a batched set of changes.
+type Client struct {
+	client     mcp.AggregatedMeshConfigServiceClient
+	stream     mcp.AggregatedMeshConfigService_StreamAggregatedResourcesClient
+	state      map[string]*perTypeState
+	clientInfo *mcp.Client
+	updater    Updater
+}
+
+// New creates a new instance of the MCP client for the specified message types.
+func New(mcpClient mcp.AggregatedMeshConfigServiceClient, supportedMessageNames []string, updater Updater, id string, metadata map[string]string) *Client { // nolint: lll
+	clientInfo := &mcp.Client{
+		Id: id,
+		Metadata: &types.Struct{
+			Fields: map[string]*types.Value{},
+		},
+	}
+	for k, v := range metadata {
+		clientInfo.Metadata.Fields[k] = &types.Value{
+			&types.Value_StringValue{v},
+		}
+	}
+
+	state := make(map[string]*perTypeState)
+	for _, messageName := range supportedMessageNames {
+		typeURL := typeURLBase + messageName
+		state[typeURL] = &perTypeState{}
+	}
+
+	return &Client{
+		client:     mcpClient,
+		state:      state,
+		clientInfo: clientInfo,
+		updater:    updater,
+	}
+}
+
+// Probe point for test code to determine when the client is finished processing responses.
+var handleResponseDoneProbe = func() {}
+
+func (c *Client) handleResponse(response *mcp.MeshConfigResponse) error {
+	if handleResponseDoneProbe != nil {
+		defer handleResponseDoneProbe()
+	}
+
+	sendNACKRequest := func(version string, err error) error {
+		log.Errorf("MCP: sending NACK for version=%v nonce=%v: error=%q", version, response.Nonce, err)
+
+		errorDetails, _ := status.FromError(err)
+		req := &mcp.MeshConfigRequest{
+			Client:        c.clientInfo,
+			TypeUrl:       response.TypeUrl,
+			VersionInfo:   version,
+			ResponseNonce: response.Nonce,
+			ErrorDetail:   errorDetails.Proto(),
+		}
+		return c.stream.Send(req)
+	}
+
+	state, ok := c.state[response.TypeUrl]
+	if !ok {
+		errDetails := status.Error(codes.Unimplemented, "unsupported type_url: %v")
+		return sendNACKRequest("", errDetails)
+	}
+
+	responseMessageName := response.TypeUrl
+	// extract the message name from the fully qualified type_url.
+	if slash := strings.LastIndex(response.TypeUrl, "/"); slash >= 0 {
+		responseMessageName = response.TypeUrl[slash+1:]
+	}
+
+	change := &Change{
+		MessageName: responseMessageName,
+		Objects:     make([]*Object, 0, len(response.Envelopes)),
+	}
+	for _, envelope := range response.Envelopes {
+		var message proto.Message
+
+		if proto.MessageType(responseMessageName) != nil {
```

Comment:
A comment here would be helpful. Also, this will "fail" for cases of generic protos (i.e. Struct, Any, Empty). Unlikely to happen as we probably won't transfer those types directly, but nevertheless worth a comment.